File: java/leetcode/PartitionLabels.java
//Leetcode problem 763 Parition Labels
//Solution written by Xuqiang Fang on 7 April, 2018 
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Iterator;
class Solution{
    public List<Integer> partitionLabels(String S) {
        List<Integer> list = new ArrayList<>();
        if(S == null)
            return list;

        HashMap<Character, Value> map = new HashMap<>();
        List<Character> order = new ArrayList<>();
        for(int i=0; i<S.length(); i++){
            char c = S.charAt(i);
            if(!map.containsKey(c)){
                map.put(c, new Value(i,i));
                order.add(c);
            }
            else{
                map.get(c).last = i;
            }
        }
        int first = 0;
        int last = map.get(order.get(0)).last;
        for(int i=1; i<order.size(); i++){
            char c = order.get(i);
            if(map.get(c).first < last){
                last = last > map.get(c).last ? last : map.get(c).last;
            }
            else{
                list.add(last - first + 1);
                first = map.get(c).first;
                last = map.get(c).last;
            }
        }
        list.add(S.length()-first);
        return list;
    }	
    class Value{
        int first;
        int last;
        public Value(int x, int y){
            first = x;
            last = y;
        }
    }
}

public class PartitionLabels{
	public static void main(String[] args){
		Solution s = new Solution();
        String S = "ababcbacadefegdehijhklijjfdkkfdmdfdfdjkfdoooopppqqrrssttn";
        System.out.println(s.partitionLabels(S));
	}
}

File: python/leetcode/PartitionLabels.py
'''
Leetcode problem No 763 Partition Labels
Solution written by Xuqiang Fang on 18 July, 2018
'''

class Solution:
    def partitionLabels(self, S):
        """
        :type S: str
        :rtype: List[int]
        """
        index = {}
        for i in range(len(S)):
            if S[i] not in index:
                index[S[i]] = i
            else:
                index[S[i]] = i
        print(index)
        curr = 0; prev = 0
        ans = []
        for i in range(len(S)):
            s = S[i]
            #print(curr)
            if index[s] == curr and i == curr:
                #print(s, i)
                ans.append(i-prev+1)
                prev = i+1
                curr = prev
            else:
                curr = max(curr, index[s])
        if prev < len(S):
            ans.append(len(S)-prev)
        return ans
                    

def main():
    s = Solution()
    S = "ababcbacadefegdehijhklij"
    print(s.partitionLabels(S))
    S = "eaaaabaaec"
    print(s.partitionLabels(S))
    S = "vhaagbqkaq"
    print(s.partitionLabels(S))

if __name__ == '__main__':
    main()
